Feature repository tags

magley opened the pull request:
Closes #36 

Tag data are kept in the database. Client app lists tags of any repo. You can filter by name, sort by last push or name, and it's paginated.

Deleting tags not implemented, I'd suggest making that its own issue (separate from deleting a repository). Also, tags don't display their manifest and size, as that could also be its own issue (it requires changing the database model and integration with the docker distribution API), assuming we need it.

To test manually, I'd push tags as `user1` onto `org1/dsa`.

magley committed:
feat: Create `Tag` model

magley committed:
feat: Add repository layer for tags

magley committed:
feat: Add tag service layer

magley committed:
feat: Tag can have null name

This is for the "default" tag that *must* be present once you push a repository.

magley committed:
feat: Docker registry will insert tag to DB on successful repo push

magley committed:
feat: Add tag controller with `find all by repo` endpoint

magley committed:
feat: Implement "advanced" tag filtering

magley committed:
feat: Create tag filter fields in web app

magley committed:
feat: Create web app `TagService`

magley committed:
feat: Repo tags component properly sorts by selected field

magley committed:
feat: Add tag details to web app

magley committed:
feat: Tag filtering returns total count for pagination

magley committed:
fix: Count elements before ordering

If you have an aggregation function, the fields which you sort by must also be in group by.

magley committed:
feat: Web app supports pagination for tags

magley committed:
feat: Add overlay popup for dates

magley committed:
fix: Properly filter tag by repository

Forgot to join tables. I'm surprised this didn't raise an exception.

magley committed:
test: Write unit tests

I skipped the `get_tags_by_` methods that just delegate to the repo layer.

magley committed:
test: Write integration tests

Vasilijez commented on client/src/components/RepoTags.tsx:
```diff
@@ -1,15 +1,143 @@
+                </Button>
+            </div>
+
+            {/* Results */}
+            {
+                tags.map((tag) => (
+                    <Col key={tag.id} xs={12}>
+                        <Card className="m-3">
+                            <Card.Body>
+                                {/* Tag title */}
+                                <Card.Title style={{ fontSize: '0.8rem' }}>Tag</Card.Title>
+
```
What do you think about adding the author of the latest push? If it feels like too much for this feature, it could be implemented as a standalone feature instead.

magley replied:
Backend: [022e772](https://github.com/magley/mocker-hub/pull/37/commits/022e77297ea2f1ab5f91b43bb39aefb53764f9ad)

Frontend: [4679ce8](https://github.com/magley/mocker-hub/pull/37/commits/4679ce8dbd6984071ebcb0d1cf5e236b10e611e5)

Vasilijez commented on client/src/components/RepoTags.tsx:
```diff
@@ -1,15 +1,143 @@
+    const orderByOptions = [
+        "Newest", "Oldest", "Name A-Z", "Name Z-A"
+    ]
+    const [filterText, setFilterText] = useState<string>("");
```
Perhaps we could support trimming whitespace from the beginning and end of the search text.

magley replied:
[2ded4fa](https://github.com/magley/mocker-hub/pull/37/commits/2ded4fa874ff787df12171b147f69b7980a82495)

I've done it server-side.

Vasilijez commented on server/app/api/tags/tag_service.py:
```diff
@@ -0,0 +1,99 @@
+        
+        # Access control.
+        
+        if not self.access_control_service.has_write_access(user_id, tag.repository_id):
+            raise AccessDeniedException(f"User {user_id} cannot remove tag {tag.name} to repo {tag.repository.canonical_name}")
```
Typo: `remove` should be `update`.

magley replied:
[fc504ac](https://github.com/magley/mocker-hub/pull/37/commits/fc504ac169e10d100f6c2ee8699897be5d1c1aa4)

Vasilijez commented on server/app/api/tags/tag_service.py:
```diff
@@ -0,0 +1,99 @@
+        
+        # Check if tag already exists for the repo.
+
+        existing_tag = self.tag_repo.find_by_name_and_repo_id(tag_name, repo_id)
+        if existing_tag:
+            raise UserException(f"Tag '{tag_name}' already exists for repository {repo.name}.")
```
It’s a redundant check and possibly dead code, but I assume you left it because it is a utility function. If that’s the case, it’s fine to leave it as is.

magley replied:
I'll leave the reply for this comment, the one below (https://github.com/magley/mocker-hub/pull/37#discussion_r1929626734) and your review comment (https://github.com/magley/mocker-hub/pull/37#pullrequestreview-2574100542).

There was two ways to fix this: either classic orchestration based error handling done in distributed systems or design by contract. I chose the second one. Basically, we assume that if a `image push` notification is sent from Distribution to our server, then that means that the repository exists, the user exists, the user has proper access to the repo etc. So, in other words, creating the tag in the DB _must_ succeed (unless the DB crashed). Take a look at the diagram below:

![Untitled Diagram drawio](https://github.com/user-attachments/assets/b3d0ced6-d819-46c1-aeea-f7ebc4c77ac5)

To handle the infinite loop, I simply catch all exceptions in the `/notifications` endpoint and log them. The endpoint will always return 200 OK.
Once you implement deleting tags from distribution, you could potentially delete the tag whenever an exception is thrown in that endpoint. That way we maintain consistency.

[f110964](https://github.com/magley/mocker-hub/pull/37/commits/f110964a362450b4ba8f5af98ced65d59d25f867)

Vasilijez commented on server/app/api/tags/tag_service.py:
```diff
@@ -0,0 +1,99 @@
+    def _touch_tag(self, user_id: int | None, tag_id: int) -> Tag:
+        # Fetch the tag.
+
+        tag = self.tag_repo.find_by_id(tag_id)
+        if tag is None:
+            raise NotFoundException(Tag, tag_id)
```
It’s a redundant check and possibly dead code, but I assume you left it because it is a utility function. If that’s the case, it’s fine to leave it as is.

Vasilijez reviewed (commented):
If I trigger an exception in `tag_service.py`, both the Distribution and Backend containers start an infinite loop. Additionally, no error message is shown in the shell where the Docker command was initiated.

magley committed:
fix: Fix typo `remove` -> `update`

magley committed:
feat: Trim whitespace in tag search queries

magley committed:
feat: Server keeps track of who last pushed a tag

magley committed:
feat: Show user who last pushed a tag in web app

magley committed:
feat: Handle exceptions when pushing tags

Basically, we should assume that `create_or_touch` (now renamed to `on_push`) never fails because it's called by Distribution AFTER it pushes the image data, which implies that access control passed.

magley committed:
test: Remove obsolete test cases
